src/api/play.rs

```diff
@@ -0,0 +1,219 @@
+use actix_identity::Identity;
+use actix_web::{web, HttpResponse};
+use aes_gcm::{
+    aead::{generic_array::GenericArray, rand_core::RngCore, Aead, OsRng},
+    Aes256Gcm, KeyInit, Nonce,
+};
+use base64::{engine::general_purpose::STANDARD, Engine};
+use octocrab::Octocrab;
+use once_cell::sync::Lazy;
+use percent_encoding::{utf8_percent_encode, NON_ALPHANUMERIC};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+use crate::{
+    api::error::{ApiError, PlaygroundError},
+    db::{
+        self,
+        model::Playground,
+        play::{create_playground, flag_playground},
+        Pool,
+    },
+    settings::SETTINGS,
+};
+
+const FILENAME: &str = "playground.html";
+const DESCRIPTION: &str = "Code shared from the MDN Playground";
+
+#[derive(Deserialize, Serialize)]
+pub struct PlayCode {
+    code: String,
+}
+#[derive(Serialize)]
+pub struct PlaySaveResponse {
+    id: String,
+}
+
+pub struct Gist {
+    pub id: String,
+    pub url: Url,
+    pub code: String,
+}
+
+#[derive(Deserialize)]
+pub struct PlayFlagRequest {
+    id: String,
+    reason: Option<String>,
+}
+
+pub(crate) const NONCE_LEN: usize = 12;
+static CIPHER: Lazy<Option<Aes256Gcm>> = Lazy::new(|| {
+    SETTINGS
+        .playground
+        .as_ref()
+        .map(|playground| Aes256Gcm::new(GenericArray::from_slice(&playground.crypt_key)))
+});
+
+fn encrypt(gist_id: &str) -> Result<String, PlaygroundError> {
+    if let Some(cipher) = &*CIPHER {
+        let mut nonce = vec![0; NONCE_LEN];
+        OsRng.fill_bytes(&mut nonce);
+        let nonce = Nonce::from_slice(&nonce);
+        let mut data = cipher.encrypt(nonce, gist_id.as_bytes())?;
+        data.extend_from_slice(nonce.as_slice());
+
+        Ok(STANDARD.encode(data))
+    } else {
+        Err(PlaygroundError::SettingsError)
+    }
+}
+
+fn decrypt(encoded: &str) -> Result<String, PlaygroundError> {
+    if let Some(cipher) = &*CIPHER {
+        let data = STANDARD.decode(encoded)?;
+        let (enc, nonce) = data.split_at(data.len() - NONCE_LEN);
+        let nonce = Nonce::from_slice(nonce);
+        let data = cipher.decrypt(nonce, enc)?;
+
+        Ok(String::from_utf8(data)?)
+    } else {
+        Err(PlaygroundError::SettingsError)
+    }
+}
+
+impl From<octocrab::models::gists::Gist> for Gist {
+    fn from(other: octocrab::models::gists::Gist) -> Self {
+        let mut files: Vec<_> = other
+            .files
+            .into_iter()
+            .map(|(name, file)| (name, file.content))
+            .collect();
+
+        files.sort_by(|(name1, _), (name2, _)| name1.cmp(name2));
+
+        let code = match files.len() {
+            0 | 1 => files
+                .into_iter()
+                .map(|(_, content)| content.unwrap_or_default())
+                .collect(),
+            _ => files
+                .into_iter()
+                .map(|(name, content)| format!("// {}\n{}\n\n", name, content.unwrap_or_default()))
+                .collect(),
```

Comment:
I'm a little confused why this branch exists - since we create the gists, shouldn't they always have exactly 1 file in them? And doesn't someone need to know our secret to be able to load random gists, which may contain more than one file?

Reply:
yep, I already have this in pending changes, good catch